Domains: Your Domains

frosty opened the pull request:
This is the first PR for bringing Domains to WPiOS. I wanted to merge early and often, particularly because this contains some Core Data model changes so I didn't want to get far behind. It contains a first stab at adding Your Domains functionality. It just displays a list of domains associated with a selected site:

* 'Domains' menu item added to blog details, feature flagged to debug
* Tapping it will fetch and display a list of domains associated with your site.
* The top two items on this screen (find a new domain / connect domain) are nonfunctional / unsettled and are just placeholders.
* The Your Domains cells should be styled as per the design (note that the design is still using Open Sans and some of the spacing is off from our default header sizes, etc):

![domain-your-domains](https://cloud.githubusercontent.com/assets/4780/15399917/6e368b84-1de1-11e6-9c42-9a4c5e4f52cf.png)

* Tapping a domain currently does nothing.
* Domains are cached in Core Data. After adding this functionality, I'm not actually sure if we need to store them? We could fetch them every time we need to display them, I guess. But equally, it can't hurt to have them there?
* The "Primary" domain marker should be displayed against the correct domain and sorted to the top of the list.
* The domain type text should be correct for each domain (see below).

**To test:** Explore this section of the app for various sites that have different types of domains. There are 4 different domain types (rules taken from Calypso):

* Registered Domain: Registered with us, JSON has the `has_registration` flag set to true.
* Site Redirect: JSON has `type` of `redirect`.
* Included with Site: A standard *.wordpress.com domain – JSON has the `wpcom_domain` flag set to true.
* Mapped Domain: Anything else.

Needs review: @astralbodies 
Thank you!

astralbodies commented:
There may be a bad merge in the project file. I'm getting this warning from CocoaPods upon an install:
```
[!] `<PBXGroup path=`Models` UUID=`2F706A870DFB229B00B43086`>` attempted to initialize an object with an unknown UUID. `5D333A561AA7A9E200DA295F` for attribute: `children`. This can be the result of a merge and  the unknown UUID is being discarded.
```

astralbodies commented on WordPress/Classes/Models/Domain.swift:
```diff
@@ -0,0 +1,66 @@
+        case .Mapped:
+            return NSLocalizedString("Mapped Domain", comment: "Describes a domain that was mapped to WordPress.com, but registered elsewhere")
+        case .SiteRedirect:
+        return NSLocalizedString("Site Redirect", comment: "Describes a site redirect domain")
```
Indent is inconsistent with above indenting.

astralbodies commented on WordPress/Classes/Networking/DomainsRemote.swift:
```diff
@@ -0,0 +1,155 @@
+import Foundation
+
+class DomainsRemote: ServiceRemoteREST {
```
I believe we're still following the pattern of `[Entity]ServiceRemote` so you might want to refactor this class name.

astralbodies commented on WordPress/Classes/Networking/DomainsRemote.swift:
```diff
@@ -0,0 +1,155 @@
+    return .Mapped
+}
+
+struct DomainsService {
```
Not sure what @koke thinks but we may wish to keep the remotes still separate from the local services in terms of source files.

frosty replied:
Sure, I can do this. I started them off in the same file for slightly easier development, but it totally makes sense to split them out now.

astralbodies commented on WordPress/Classes/Networking/DomainsRemote.swift:
```diff
@@ -0,0 +1,155 @@
+
+    private let context = ContextManager.sharedInstance().mainContext
+
+    init(blog: Blog) {
```
All of the local services should be state-free with the exception of the managed object context being the only initialization parameter. If you look at (hopefully all) of the other local Core Data services each method takes all of the parameters it needs to function. There may be methods in this service to query across Blogs so it makes sense to not lock it down to one and not retain some state across methods.

frosty replied:
Ah, okay, thanks – I actually hadn't realised that services weren't supposed to have any state. I was sort of basing things on Plans (blaming @koke :grin:), where the service has a `store` (for talking to StoreKit) and a `remote`. So I should pass all of these things in to each method?

frosty replied:
I've updated the service so that the only state it retains is the `context` (private constant) and the `remote` (I noticed that `NotificationsService` seemed to be doing the same thing). The context is now passed into the initializer, and the `siteID` is now passed in as required, instead of retaining a `blog` reference. 

Is this okay, or would you like me to take the remote out as well?

astralbodies commented on WordPress/Classes/Models/Domain.swift:
```diff
@@ -0,0 +1,66 @@
+}
+
+class ManagedDomain: NSManagedObject {
+    @NSManaged var domain: String
```
Would domainName here be a bit less absolute? I was poking through code later in the local service and wasn't sure what "domain" was in the fetch request at first. 😄

frosty replied:
Definitely sounds better to me :+1:

astralbodies commented on WordPress/Classes/Networking/DomainsRemote.swift:
```diff
@@ -0,0 +1,155 @@
+
+    private func blogDomains() -> [Domain] {
+        let request = NSFetchRequest(entityName: Domain.entityName)
+        request.predicate = NSPredicate(format: "%K == %@", "blog", blog)
```
Any specific reason the property here is a parameter?

frosty replied:
Copy pasta :) I'll add it straight into the string here – although in previous projects I've preferred using a parameter and an `Entity.entityName` property to avoid the chance of typos.

frosty replied:
In fact, I've gone all-in and switched out all literals that referred to attribute names with references to static constants I've put on `ManagedDomain`. e.g.: 

`NSPredicate(format: "%K == %@", ManagedDomain.Relationships.blog, blog)`

Helps with things like renaming attributes and making sure you've done it everywhere :)

astralbodies commented:
Also don't kill me but could you write a minimal set of unit tests for the services?

frosty commented:
Regarding the CocoaPods warning – looks like the project had a dangling reference to an old file that's been removed from the project (`WPPostContentViewProvider.h`, removed here: https://github.com/wordpress-mobile/WordPress-iOS/pull/5392). Perhaps I accidentally kept that line when resolving a merge. I've removed it, and the CocoaPods error has now gone.

I think I've now addressed all your other comments, including adding some tests – would you care to take a look again @astralbodies? Please let me know if I've missed anything, or if you want me to take a look at anything else! Thanks! :)

astralbodies commented:
@frosty Looks like there are merge conflicts now 😕

frosty commented:
Okay, conflicts fixed – sorry about that @astralbodies! We regret the error.

astralbodies commented:
Nice job! If the tests pass, :shipit: 

![high19](https://cloud.githubusercontent.com/assets/373903/15515929/08f61734-21b6-11e6-9a3b-0ecb1583f3d7.gif)

astralbodies commented:
Woot! :shipit: for sure 😄

frosty commented:
Woo, thanks for the review @astralbodies!
